twiist/provider/provider.go

```diff
@@ -68,65 +160,150 @@ func (p *Provider) OnCreate(ctx context.Context, userID string, providerSessionI
+	var exists bool
+	for _, dataSetID := range dataSetIDs {
+		if dataSetID == *dataSet.ID {
+			exists = true
+			break
 		}
-		if dataSourceID, ok := task.Data["dataSourceId"].(string); ok && dataSourceID != "" {
-			update := dataSource.NewUpdate()
-			update.State = pointer.FromString(dataSource.StateDisconnected)
-			_, err = p.dataSourceClient.Update(ctx, dataSourceID, nil, update)
-			if err != nil {
-				logger.WithError(err).WithField("dataSourceId", dataSourceID).Error("unable to update data source after deleting provider session")
-			}
+	}
+
+	// Only append the Data Set ID if it's unique
+	if !exists {
+		dataSetIDs = append(dataSetIDs, *dataSet.ID)
+	}
+
+	update := dataSource.Update{
+		DataSetIDs:         pointer.FromStringArray(dataSetIDs),
+		ProviderExternalID: providerSession.ExternalID,
+		ProviderSessionID:  pointer.FromString(providerSession.ID),
+		State:              pointer.FromString(dataSource.StateConnected),
+	}
+
+	// The external id of the data source and provider session is set to the Tidepool Link ID.
+	// The Tidepool Link ID changes every time a user re-links their account. We also need to
+	// keep track of the twiist user id which is different from the Tidepool Link ID, in case
+	// we need to support backfilling of historical data and a user has more than one twiist
+	// accounts. The twiist user id can be found in the id token.
+	claims, err := p.getClaimsFromIDToken(ctx, providerSession.OAuthToken)
+	if err != nil {
+		return errors.Wrap(err, "unable to get claims from id token")
+	}
+	if claims.Subject != "" {
+		update.Metadata = source.Metadata
+		if update.Metadata == nil {
+			update.Metadata = make(map[string]any)
 		}
+		update.Metadata[MetadataKeyExternalSubjectID] = claims.Subject
 	}
+
+	_, err = p.dataSourceClient.Update(ctx, *source.ID, nil, &update)
+	if err != nil {
+		return errors.Wrap(err, "unable to update source with data set id")
+	}
+
 	return nil
 }
+
+func (p *Provider) findDataSet(ctx context.Context, source *dataSource.Source) (*data.DataSet, error) {
+	if source.DataSetIDs != nil {
+		for index := len(*source.DataSetIDs) - 1; index >= 0; index-- {
+			if dataSet, err := p.dataClient.GetDataSet(ctx, (*source.DataSetIDs)[index]); err != nil {
+				return nil, errors.Wrap(err, "unable to get data set")
+			} else if dataSet != nil {
+				return dataSet, nil
+			}
+		}
+	}
+	return nil, nil
+}
+
+func (p *Provider) createDataSet(ctx context.Context, source *dataSource.Source) (*data.DataSet, error) {
+	dataSetCreate := data.NewDataSetCreate()
+	dataSetCreate.Client = &data.DataSetClient{
+		Name:    pointer.FromString(twiist.DataSetClientName),
+		Version: pointer.FromString(twiist.DataSetClientVersion),
+	}
+	dataSetCreate.DataSetType = pointer.FromString(data.DataSetTypeContinuous)
+	dataSetCreate.Deduplicator = data.NewDeduplicatorDescriptor()
+	dataSetCreate.Deduplicator.Name = pointer.FromString(dataDeduplicatorDeduplicator.DataSetDeleteOriginName)
```

Comment:
This needs to be the new `dataDeduplicatorDeduplicator.DataSetDeleteOriginOlderName` once it is merged. (Note "Older")